File: python/213_House_Robber_II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        length = len(nums)
        if length == 0:
            return 0
        if length == 1:
            return nums[0]

        def my_rob(nums):
            cur, pre = 0, 0
            for num in nums:
                cur, pre = max(pre + num, cur), cur
            return cur
        
        return max(my_rob(nums[:-1]),my_rob(nums[1:]))

File: java/213_House_Robber_II.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        int[] G = new int[n+1];
        int[] N = new int[n+1];

        G[1] = nums[0];
        for(int i = 2 ; i <= n ; i++){
            if(i != n){
                G[i] = Math.max(G[i-1],G[i-2]+nums[i-1]);
            }else{
                G[i] = Math.max(G[i-1],G[i-2]);
            }
        }

        for(int i = 2 ; i <= n ; i++){
            N[i] = Math.max(N[i-1],N[i-2]+nums[i-1]);
        }

        return Math.max(G[n],N[n]);
    }
}